libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * @file Inliner.cpp
+ * Inlines small code snippets by replacing JUMP with a copy of the code jumped to.
+ */
+
+#include <libevmasm/Inliner.h>
+
+#include <libevmasm/AssemblyItem.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/slice.hpp>
+
+#include <optional>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::evmasm;
+
+bool Inliner::isInlineCandidate(InlinableBlock const& _block) const
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+	// Always try to inline if there is at most one call to the block.
+	if (_block.pushTagCount == 1)
+		return true;
```

Comment:
The problem the way this is written for now is that if we're unlucky, we'll already have the one block pushing this tag ten times before this is called... maybe that's worth accounting for - should be possible by tracking all ``PushTag``s we've already inlined...

Reply:
Should be fine now.